src/controllers/orderController.js

```diff
@@ -39,4 +42,33 @@ const createOrder = async (req, res) => {
     .json({ clientSecret: paymentIntent.client_secret, order });
 };
 
-module.exports = { createOrder };
+const getAllOrders = async (req, res) => {
+  const orders = await Order.find({});
+  res.status(StatusCodes.OK).json({ orders, count: orders.length });
+};
+
+const getSingleOrder = async (req, res) => {
+  const { id: orderId } = req.params;
+  const order = await Order.findOne({ _id: orderId });
+  if (!order) {
+    throw new CustomError.NotFoundError(`No order with id ${orderId}`);
+  }
+
+  checkPermissions(req.user, order.user);
+
+  res.status(StatusCodes.OK).json({ order });
+};
+
+const deleteOrder = async (req, res) => {
+  const { id: orderId } = req.params;
+  const order = await Order.findOneAndDelete({ _id: orderId });
+
+  if (!order) {
+    throw new CustomError.NotFoundError(`No order with id ${orderId}`);
+  }
+  checkPermissions(req.user, order.user);
```

Comment:
At this point, the order has already been deleted if it was found on line 64, so checking if this was the right user here will have no effect. If it was the wrong user, it will throw a Unauhtorized error, but we'll still have already deleted the order.

Two options:

1.  split the "find order" and "split order" steps
2.  use a transaction that will rollback if there are any errors (https://mongoosejs.com/docs/transactions.html)

```javascript
// Option 1
const order = await Order.findById(orderId)
if (!order) {
  throw new NotFoundError(`No order with id ${orderId}`);
}
checkPermissions(req.user, order.user)

await Order.findByIdAndDelete(orderId); // or more simply can just call delete on the documnent we've already fetched:  await order.remove()

res.status(StatusCodes.OK).json({ msg: 'Success! Order was deleted' })

// Option 2
const session = await Order.startSession // start a MongoDb client session
await session.withTransaction(async () => {
  // everything in this block is in a transaction, which means it will be undone if there is an error thrown
  const order = await Order.findOneAndDelete({ _id: orderId });

  if (!order) {
    throw new CustomError.NotFoundError(`No order with id ${orderId}`);
  }
  checkPermissions(req.user, order.user); // if this throws an "UnauthorizedError" the delete step above will be reverted
});

// assuming we get here, that means we haven't thrown an error, so we can return suiccess response:
res.status(StatusCodes.OK).json({msg: "Success! Order was deleted" })'
```

Reply:
@akosasante thank you  so much for showing new approaches!!! Appreciate it ❤️